### PR TITLE
[WIP] Add std.meta.apply

### PR DESCRIPTION
**Pull request by PetarKirov**
TODO:

- [ ] add more tests which properly motivate this addition
- [ ] add changelog entry

### CODE REVIEW

**Comment by schveiguy** [comment]
Doesn't `Instantiate` already do something similar to this? Would it not be better to simply extend `Instantiate` to do things you need?

**Comment by JackStouffer** [comment]
bike-sheding: this should be named `Apply` to conform with the naming conventions of the rest of std.meta

**Comment by ZombineDev** [comment]
@schveiguy yes, unfortunately `Instantiate` is similar, but not quite as useful, I need to add more tests which properly motivate this addition.

@JackStouffer I considered this, but it wasn't a clear call since it can be used with values, where `apply` would be more appropriate and with templates, where `Apply` would be better. However the most important point is that `apply` / `Apply` combines both behaviours, which brings a tremendous simplification in generic code.

File: std/meta.d [hunk 1]
```diff
@@ -1210,6 +1211,147 @@ template aliasSeqOf(alias range)
     }
 }
 
+/**
+ * Applies the sequence of arguments `args` to the function or template `func`
+ * either as function or template arguments, at compile-time or at run-time
+ * (if CTFE is not possible).
```
Comment (andralex) [inline std/meta.d]: The introduction paragraph should answer the obvious question "What's wrong with `func(args)`?"

File: std/meta.d [hunk 2]
```diff
@@ -1210,6 +1211,147 @@ template aliasSeqOf(alias range)
...
+ *
+ * `apply` is most likely to be found used as a building block in higher-order
+ * templates and $(LREF AliasSeq) algorithms like $(LREF staticMap) which need
+ * to call or instantiate (as appropriate) functions or templates, passed as
+ * parameters.
+ *
+ * Params:
+ *      func  = function or template to call or instantiate
+ *      args = template / function arguments to `apply` to `fun`
```
Comment (andralex) [inline std/meta.d]: no spaces around "/" in text

File: std/meta.d [hunk 3]
```diff
@@ -1210,6 +1211,147 @@ template aliasSeqOf(alias range)
...
+    int add(int a, int b) { return a + b; }
+    enum int a = 2, b = 3;
+    enum int sum = apply!(add, a, b);
+    static assert(sum == 5);
+
+    int x = 2, y = 5;
+    int z = apply!((a, b) => a + b, x, y);
+    assert(z == 7);
+
+    // TODO: Show higher-order template example
```
Comment (andralex) [inline std/meta.d]: This can't be part of the published documentation.

**Comment by LightBender** [comment]
PR closed as stalled. If you wish to resurrect it you are welcome to do so. @jmdavis If this has any relevance to your Phobos 3 work let us know.

**Pull request closed without merging**